### ISSUE TITLE
Unhandled exception in method ReadLink(string path)

### ISSUE THREAD

**Issue by MaksimAnishko** [AvaloniaUI/Avalonia#3600]
Hello.

Unhandled exception throw in Avalonia.FreeDesktop.NativeMethods.**ReadLink(String path)**.

> "The output byte buffer is too small to contain the encoded data, encoding 'Unicode (UTF-8)' fallback 'System.Text.EncoderReplacementFallback'."

**Input parameter**: "/dev/disk/by-label/Зарезервировано\\x20системой"

Maybe instead of
`var symlinkMaxSize = Encoding.ASCII.GetMaxByteCount(path.Length);`

use 

`var symlinkMaxSize = Encoding.UTF8.GetMaxByteCount(path.Length);`

Thanks.

**Comment by kekekeks** [comment]
1) PATH_MAX is in byte len, not character len.
2) it attempts to use ASCII encoding for UTF-8 paths for some weird reason.

So the entire thing is broken with non-ASCII characters.


@jmacato

**Comment by kekekeks** [comment]
Also, readlink resolves relative symlinks using the current working directory.

**Comment by kekekeks** [comment]
@jmacato you should use [realpath](man7.org/linux/man-pages/man3/realpath.3.html) instead since it already implements proper path resolving logic.

**Comment by jmacato** [comment]
Yeah, i need to convert the ASCII byte formatter into UTF8; As well as use realpath instead of reallink

**Comment by kekekeks** [comment]
Just declare it as

```cs
        [DllImport("libc", SetLastError = true)]
        private static IntPtr realpath(string path, IntPtr unused);
        [DllImport("libc"]
        private static void free(IntPtr mem);
```

```cs
var resolvedPtr = realpath(pathString, IntPtr.Zero);
if(resolvedPtr == IntPtr.Zero)
    throw new Win32Exception();
var rv = Marshal.PtrToStringAnsi(resolvedPtr);
free(resolvedPtr)
return rv;

```

Marshaller will do all of UTF8 conversions automatically without extra manual work that you were doing incorrectly anyway

**Comment by MaksimAnishko** [comment]
The issue no longer reproduces.
